lumicks/pylake/kymotracker/detail/binding_times.py

```diff
@@ -0,0 +1,377 @@
+        hist_kwargs : dict
+            dictionary of plotting kwargs applied to histogram
+        span_kwargs : dict
+            dictionary of plotting kwargs applied to the patch indicating the area
+            spanned by the confidence intervals
+        line_kwargs : dict
+            dictionary of plotting kwargs applied to the line indicating the
+            distribution mean
+        """
+        hist_kwargs = {"facecolor": "#c5c5c5", "edgecolor": "#888888", **hist_kwargs}
+        span_kwargs = {"facecolor": "tab:red", "alpha": 0.3, **span_kwargs}
+        line_kwargs = {"color": "k", **line_kwargs}
+
+        for j, (key, unit) in enumerate(zip(("amplitude", "lifetime"), ("", "sec"))):
+            data = getattr(self, f"{key}_distributions")[component]
+            plt.subplot(1, 2, j + 1)
+            plt.hist(data, bins=n_bins, **hist_kwargs)
+
+            mean, (lower, upper) = self.calculate_stats(key, component, alpha)
+            plt.axvspan(lower, upper, **span_kwargs)
+            plt.axvline(mean, **line_kwargs)
+            plt.xlabel(key)
+            plt.ylabel("counts")
+            plt.title(f"{mean:0.2g} ({lower:0.2g}, {upper:0.2g}) {unit}")
+
+        plt.tight_layout()
+
+
+@dataclass(frozen=True)
+class BindingDwelltimes:
+    """Results of exponential mixture model optimization for binding dwelltimes.
+
+    This class is returned from `_kinetic_mle_optimize()` and should not be
+    constructed manually.
+
+    Attributes
+    ----------
+    n_components : int
+        number of components in the model.
+    dwelltimes_sec : np.ndarray
+        observations on which the model was trained.
+    observations_limits : tuple
+        tuple of (`min`, `max`) values of the experimental observation time.
+        Generally these correspond to the frame/line time and the experiment
+        length, respectively.
+    _parameters : np.ndarray
+        optimized parameters in the order [amplitudes, lifetimes]
+    log_likelihood : float
+        log likelihood of the trained model
+    bootstrap : BindingDwelltimesBootstrap
+        object containing information about the bootstrapping analysis.
+    """
+
+    n_components: int
+    dwelltimes_sec: np.ndarray = field(repr=False)
+    observation_limits: list = field(repr=False)
+    _parameters: np.ndarray = field(repr=False)
+    log_likelihood: float
+    bootstrap: BindingDwelltimesBootstrap = field(
+        default_factory=BindingDwelltimesBootstrap, repr=False
+    )
+
+    @property
+    def amplitudes(self):
+        """Fractional amplitude of each model component"""
+        return self._parameters[: self.n_components]
+
+    @property
+    def lifetimes(self):
+        """Lifetime parameter (in seconds) of each model component."""
+        return self._parameters[self.n_components :]
+
+    @property
+    def aic(self):
+        """Akaike Information Criterion."""
+        k = 2 * self.n_components  # number of parameters
```

Comment:
As far as I know, the Akaike Information Criterion has as an assumption that the parameter estimation problem is not subject to constraints. This one is.

I'm wondering if the number of parameters is correct here. I think every active constraint reduces it by one parameter (since one can rephrase the model in an unconstrained problem with one fewer parameter in this case), but this would have to be verified.

Do people typically use aic for model selection in these models?

Reply:
yes, this is fixed now; sorry forgot to mention it